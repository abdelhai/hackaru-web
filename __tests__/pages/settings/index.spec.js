import { Store } from 'vuex-mock-store';
import Factory from '@/__tests__/__setups__/factory';
import Index from '@/pages/settings/index';

describe('Index', () => {
  let factory;
  let wrapper;

  const location = jest
    .spyOn(window.location, 'assign')
    .mockImplementation(() => {});

  const $store = new Store({
    getters: {
      'auth/getEmail': 'example@example.com'
    }
  });

  beforeEach(() => {
    $store.reset();
    factory = new Factory(Index, {
      mocks: { $store }
    });
  });

  it('render correctly', () => {
    expect(factory.shallow().element).toMatchSnapshot();
  });

  describe('when submit change email form', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({
        emailForm: {
          email: 'example@example.com',
          currentPassword: 'currentPassword'
        }
      });
      wrapper.find('form.email').trigger('submit.prevent');
    });

    it('dispatch auth/changeEmail', () => {
      expect($store.dispatch).toHaveBeenCalledWith('auth/changeEmail', {
        email: 'example@example.com',
        currentPassword: 'currentPassword'
      });
    });

    it('show success toast', () => {
      expect(factory.options.mocks.$toast.success).toHaveBeenCalled();
    });
  });

  describe('when submit change email form but failed', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(false);
      wrapper = factory.shallow();
      wrapper.find('form.email').trigger('submit.prevent');
    });

    it('does not show success toast', () => {
      expect(factory.options.mocks.$toast.success).not.toHaveBeenCalled();
    });
  });

  describe('when submit change password form', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({
        passwordForm: {
          currentPassword: 'currentPassword',
          password: 'password',
          passwordConfirmation: 'passwordConfirmation'
        }
      });
      wrapper.find('form.password').trigger('submit.prevent');
    });

    it('dispatch auth/changePassword', () => {
      expect($store.dispatch).toHaveBeenCalledWith('auth/changePassword', {
        currentPassword: 'currentPassword',
        password: 'password',
        passwordConfirmation: 'passwordConfirmation'
      });
    });

    it('show success toast', () => {
      expect(factory.options.mocks.$toast.success).toHaveBeenCalled();
    });
  });

  describe('when submit change password form but failed', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(false);
      wrapper = factory.shallow();
      wrapper.find('form.password').trigger('submit.prevent');
    });

    it('does not show success toast', () => {
      expect(factory.options.mocks.$toast.success).not.toHaveBeenCalled();
    });
  });

  describe('when click logout button and confirm is true', () => {
    beforeEach(() => {
      global.confirm = () => true;
      wrapper = factory.shallow();
      wrapper.find('.logout-button').vm.$emit('click');
    });

    it('dispatch auth/logout', () => {
      expect($store.dispatch).toHaveBeenCalledWith('auth/logout');
    });

    it('send ga event', () => {
      expect(factory.options.mocks.$ga.event).toHaveBeenCalledWith(
        'auth',
        'logout'
      );
    });

    it('redirect to index', () => {
      expect(location).toHaveBeenCalledWith('/en/index');
    });
  });

  describe('when click logout button and confirm is false', () => {
    beforeEach(() => {
      global.confirm = () => false;
      wrapper = factory.shallow();
      wrapper.find('.logout-button').vm.$emit('click');
    });

    it('does not dispatch auth/logout', () => {
      expect($store.dispatch).not.toHaveBeenCalled();
    });

    it('does not redirect', () => {
      expect(location).not.toHaveBeenCalled();
    });
  });

  describe('when click deleteAccount button', () => {
    beforeEach(() => {
      wrapper = factory.shallow();
      wrapper.find('.delete-account-button').vm.$emit('click');
    });

    it('show modal', () => {
      expect(factory.options.mocks.$modal.show).toHaveBeenCalledWith(
        'deleteAccount'
      );
    });
  });

  describe('when submit delete account form and confirm is true', () => {
    beforeEach(() => {
      global.confirm = () => true;
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({
        deleteAccountForm: {
          currentPassword: 'currentPassword'
        }
      });
      wrapper.find('form.delete-account').trigger('submit.prevent');
    });

    it('dispatch auth/deleteAccount', () => {
      expect($store.dispatch).toHaveBeenCalledWith('auth/deleteAccount', {
        currentPassword: 'currentPassword'
      });
    });

    it('send ga event', () => {
      expect(factory.options.mocks.$ga.event).toHaveBeenCalledWith(
        'auth',
        'deleteAccount'
      );
    });

    it('redirect to index', () => {
      expect(location).toHaveBeenCalledWith('/en/index');
    });
  });

  describe('when submit delete account form and confirm is false', () => {
    beforeEach(() => {
      global.confirm = () => false;
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({
        deleteAccountForm: {
          currentPassword: 'currentPassword'
        }
      });
      wrapper.find('form.delete-account').trigger('submit.prevent');
    });

    it('does not dispatch auth/deleteAccount', () => {
      expect($store.dispatch).not.toHaveBeenCalled();
    });

    it('does not redirect', () => {
      expect(location).not.toHaveBeenCalled();
    });
  });

  describe('when submit delete account form but failed', () => {
    beforeEach(() => {
      global.confirm = () => true;
      $store.dispatch.mockReturnValue(false);
      wrapper = factory.shallow();
      wrapper.setData({
        deleteAccountForm: {
          currentPassword: 'currentPassword'
        }
      });
      wrapper.find('form.delete-account').trigger('submit.prevent');
    });

    it('does not redirect to index', () => {
      expect(location).not.toHaveBeenCalledWith();
    });
  });
});
