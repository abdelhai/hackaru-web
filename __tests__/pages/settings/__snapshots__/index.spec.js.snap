// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index render correctly 1`] = `
<section>
  <modal-wrapper-stub
    height="auto"
    name="deleteAccount"
  >
    <form
      class="delete-account"
    >
      <modal-header-stub>
        <h1>
          deleteAccountModal.title
        </h1>
      </modal-header-stub>
       
      <modal-item-stub>
        <icon-stub
          class="icon alert-icon"
          name="alert-triangle-icon"
        />
        
        deleteAccountModal.warning
      
      </modal-item-stub>
       
      <modal-item-stub>
        <input
          aria-label="deleteAccountModal.currentPassword"
          class="input inline"
          placeholder="deleteAccountModal.currentPassword"
          required="required"
          type="password"
        />
      </modal-item-stub>
       
      <modal-footer-stub>
        <btn-stub
          class="is-rounded is-danger"
          type="submit"
        >
          
          deleteAccountModal.delete
        
        </btn-stub>
      </modal-footer-stub>
    </form>
  </modal-wrapper-stub>
   
  <article>
    <heading-stub
      class="is-small"
    >
      <icon-stub
        class="icon"
        name="mail-icon"
      />
      
      titles.email
    
    </heading-stub>
     
    <form
      class="email"
    >
      <text-field-stub
        aria-label="email"
        class="has-border"
        placeholder="email"
        required=""
        type="email"
        value="example@example.com"
      />
       
      <text-field-stub
        aria-label="password"
        class="has-border"
        placeholder="password"
        required=""
        type="password"
        value=""
      />
       
      <btn-stub
        class="is-rounded is-primary"
        type="submit"
      >
        update
      </btn-stub>
    </form>
  </article>
   
  <article>
    <heading-stub
      class="is-small"
    >
      <icon-stub
        class="icon"
        name="lock-icon"
      />
      
      titles.password
    
    </heading-stub>
     
    <form
      class="password"
    >
      <text-field-stub
        aria-label="password"
        class="has-border"
        placeholder="password"
        required=""
        size="1"
        type="password"
        value=""
      />
       
      <text-field-stub
        aria-label="newPassword"
        class="has-border"
        placeholder="newPassword"
        required=""
        size="1"
        type="password"
        value=""
      />
       
      <text-field-stub
        aria-label="confirmNewPassword"
        class="has-border"
        placeholder="confirmNewPassword"
        required=""
        size="1"
        type="password"
        value=""
      />
       
      <btn-stub
        class="is-rounded is-primary"
        type="submit"
      >
        update
      </btn-stub>
    </form>
  </article>
   
  <article>
    <heading-stub
      class="is-small"
    >
      <icon-stub
        class="icon"
        name="globe-icon"
      />
      
      titles.language
    
    </heading-stub>
     
    <locale-select-stub />
  </article>
   
  <article>
    <heading-stub
      class="is-small"
    >
      <icon-stub
        class="icon"
        name="log-out-icon"
      />
      
      titles.logout
    
    </heading-stub>
     
    <btn-stub
      class="is-rounded logout-button is-primary"
      type="button"
    >
      
      Logout
    
    </btn-stub>
  </article>
   
  <article>
    <heading-stub
      class="is-small"
    >
      <icon-stub
        class="icon"
        name="user-x-icon"
      />
      
      titles.deleteAccount
    
    </heading-stub>
     
    <btn-stub
      class="is-rounded delete-account-button is-danger"
      type="button"
    >
      
      deleteAccount
    
    </btn-stub>
  </article>
</section>
`;
